### PR TITLE
Fixed convert function monotonic failure when string data type

### PR DESCRIPTION
**Pull request by zhang2014**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category:
- Bug Fix

Short description:

Query is failure when use toUInt64 function with primary key of string data type in where expression,
Because toUInt64 did the wrong monotonic optimization

### CODE REVIEW

File: dbms/src/Functions/FunctionsConversion.h [hunk 1]
```diff
@@ -1138,6 +1138,10 @@ struct ToIntMonotonicity
 
     static IFunction::Monotonicity get(const IDataType & type, const Field & left, const Field & right)
     {
+        /// If type is string, the function is no monotonic
+        if (WhichDataType(type).isString())
```
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsConversion.h]: 1. What about other types: FixedString for example.
Maybe better to specify condition with `isRepresentedByNumber`?

2. Conversion from String to String; from Array to Array of same type is monotonic.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
Ok but one concern remains.

File: dbms/src/Functions/FunctionsConversion.h (hunk 1 above)
Comment (zhang2014) [reply dbms/src/Functions/FunctionsConversion.h]: > What about other types: FixedString for example.
> Maybe better to specify condition with isRepresentedByNumber?

Done

> Conversion from String to String; from Array to Array of same type is monotonic.

ToIntMonotonicity is just use to `any -> Int`, So I'm not sure if it has `Array -> Array` 

And I rewrite the  test to make it testable.

**Comment by alexey-milovidov** [comment]
Let's also retain the old test,
and add your test in a new file.

**Comment by zhang2014** [comment]
Old tests are kept in:
- https://github.com/yandex/ClickHouse/pull/3870/files#diff-a485106b9dfd012a710477225331f6daR27
- https://github.com/yandex/ClickHouse/pull/3870/files#diff-a485106b9dfd012a710477225331f6daR36
- https://github.com/yandex/ClickHouse/pull/3870/files#diff-a485106b9dfd012a710477225331f6daR49

Should I restore the original SQL file ?

**Comment by alexey-milovidov** [comment]
Yes. Just in case, it's reasonable to also keep the existing case separate.
Because if both tests will be broken, the old test may be simpler for investigation.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
The test fails constantly in debug build.

**Comment by alexey-milovidov** [comment]
The business logic is broken.

```
SELECT toInt8(CounterID)
FROM test.hits 
WHERE CounterID = 23192098

┌─toInt8(CounterID)─┐
│                34 │
│                34 │
│                34 │
└───────────────────┘


SELECT CounterID
FROM test.hits 
WHERE (CounterID = 23192098) AND (toInt8(CounterID) = 34)

Ok.

0 rows in set. 
```

**Comment by alexey-milovidov** [comment]
You are reading uninitialized data.

**Comment by zhang2014** [comment]
> The business logic is broken.

Might the be added to the test case?

> You are reading uninitialized data.

Into a Null value when convert RNP?  the logic here is not very clear to me.

**Comment by alexey-milovidov** [comment]
> Into a Null value when convert RNP? the logic here is not very clear to me.

It's about `field.get<Int64>()` when field can be NULL.
Probably this error was already present.
I've rewriting it.

> Might the be added to the test case?

This was found by a test case in internal ("stateful") part of tests.

**Comment by alexey-milovidov** [comment]
#3921

**Pull request merged**